source/extensions/network/dns_resolver/getaddrinfo/getaddrinfo.cc

```diff
@@ -155,8 +157,16 @@ void GetAddrInfoDnsResolver::resolveThreadRoutine() {
         absl::MutexLock guard(&mutex_);
         pending_queries_.push_back(next_query);
         continue;
+      } else if (treat_nodata_noname_as_success &&
+                 (rc.return_value_ == EAI_NONAME || rc.return_value_ == EAI_NODATA)) {
+        // Treat NONAME and NODATA as DNS records with no results.
+        // NODATA and NONAME are typically not transient failures, so we don't expect success if
+        // the DNS query is retried.
+        // NOTE: this is also how the c-ares resolver treats NONAME and NODATA:
+        // https://github.com/envoyproxy/envoy/blob/099d85925b32ce8bf06e241ee433375a0a3d751b/source/extensions/network/dns_resolver/cares/dns_impl.h#L109-L111.
+        ENVOY_LOG(debug, "getaddrinfo no results rc={}", gai_strerror(rc.return_value_));
+        response = std::make_pair(ResolutionStatus::Success, std::list<DnsResponse>());
```

Comment:
I agree the `Success` name is confusing (added a TODO to rename to `Completed` or something similar).  c-ares treats NODATA and NONAME as "success" too: https://github.com/envoyproxy/envoy/blob/099d85925b32ce8bf06e241ee433375a0a3d751b/source/extensions/network/dns_resolver/cares/dns_impl.h#L109-L111.  The idea is really what does the DNS cache do with "success" vs. "failure".  If we have a "failure", we put re-resolution requests on a retry timer: https://github.com/envoyproxy/envoy/blob/099d85925b32ce8bf06e241ee433375a0a3d751b/source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc#L438, which uses a retry (with backoff) strategy.  If it's a "success", we use the DNS TTL for refreshes: https://github.com/envoyproxy/envoy/blob/099d85925b32ce8bf06e241ee433375a0a3d751b/source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc#L432.

Since we'd expect to retry more often in the case of "failure" until we get a successful result, the idea is that we shouldn't put queries on a "failure" refresh timer if we don't expect a different result each time we run the query, and from my discussions with @ericorth, you generally don't expect a different result when retrying NODATA and NONAME results - those aren't considered transient failures.

Reply:
For extra reference, see RFC2308, which explains the caching expectations for NXDOMAIN and NODATA results.  Most of the details in there are more relevant to implementing a full DNS resolver (client or server) where the RFC explains how to determine the cache TTL for those results, but the RFC is also useful for the general case of seeing that those return types are expected to be treated as full non-transient results to be cached just like positive with-address results.